### DISCUSSION TITLE
beta4: Clear key change is temporary. Key is back after next autokey restart.

### DISCUSSION THREAD

**Discussion by jack99999** [autokey/autokey#550] [Beta feedback]
i'm using ubuntu 20.04 with i3 and autokey-gtk 0.96.0-beta.4  

i have a couple of scripts that call the chromium browser but i wanted to use brave instead. i copied the scripts, then renamed the copies, then removed the key binding from the chromium version. when i press <Clear> for the key binding the left column of the UI does not remove the key from the list until i do a save. is that expected? feels wrong. (as would a word processor that did not show changes until a file save.)

fwiw, i see the .json file is rewritten when i save but the key definition is in the new copy of the file.

jack

**Comment by BlueDrink9** [comment]
Aha, good to know that it is the .json that ends up with the correct info. That should make this easy to fix, once I get time to look at it.

Actually wait, let me clarify: What do you mean by "the `.json` file" and what do you mean by "the new copy of the file"? Are you referring to the hidden sidecar and its new form? (e.g. `.my_phrase.json` and `my_phrase.json`)?

**Comment by BlueDrink9** [comment]
I have bisected this bug between 0.95.10 and 0.96.0-Beta.1 Sadly, there are a bunch of commits that autokey will not correctly start during, and the bug was introduced in one of them.
[bisect.log](https://github.com/autokey/autokey/files/6628264/bisect.log)

```
b5ec3fa59e781b2ec01bc89c2dec1d34cf0d20cf
93a56441ba69cefea1bd1c40c61938c20411070a
ff59912532c555b03cbff9c761b42788346dec68
5f55719ccb000e6a1758ecd3e98cea767e47bc43
74383446b1892d0dc8d52ec992f69d1138c539a7
012f51e557be22becf3d733b7da2cdb1302f684d
83a4c748f92ffa4adfe092c83609defcf6952708
2cd8d488c9599f67d5416bf25d5c0acfc7b0dc83
50c17b66fc0c311daa363d1fce23fb1bd00bf900
51ea2831dbfe09bdb0b2e0c6b23860890fd3d76c
```

I have attached the bisect log.

**Reply by BlueDrink9** [reply]
I think this is caused by 50c17b66. The [addition of line 393](https://github.com/autokey/autokey/commit/50c17b66fc0c311daa363d1fce23fb1bd00bf900#diff-a974a007e47ddd7a5211af657e180d2fdd1b6a6bd419fb923b67484f6fc1d2f4R393) brings about this behavior.

**Reply by BlueDrink9** [reply]
Although I seem to have already noticed this, [as seen by this comment](https://github.com/autokey/autokey/blob/cad9e1dd87e5297a8d1e07b4c729ff28a055527c/lib/autokey/model/abstract_hotkey.py#L48)

**Reply by BlueDrink9** [reply]
Adding the modes isn't the source of the bug however. It just unmasks the problematic behaviour.

Currently, clicking "clear" does not change how the hotkeys are persisted, but it does remove `TriggerMode.HOTKEY` from the item, which is persisted as an empty "modes" array in the json. When autokey is reloaded, the empty "modes" array means the hotkeys in that item are ignored.
With the aforementioned lines adding that triggermode back, the hotkeys that are persisted are seen and taken up as hotkeys.

It makes complete sense to add the triggermode if adding a hotkey, so those lines should stay there.
It doesn't make sense that clearing doesn't remove the previous hotkeys from the persisted `json`, so the fix will be ensuring it does.

**Reply by BlueDrink9** [reply]
Fixed in 9187259
